src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.GroupedAnalyzerActionsByAnalyzer.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal partial class AnalyzerDriver<TLanguageKindEnum> : AnalyzerDriver where TLanguageKindEnum : struct
+    {
+        private sealed class GroupedAnalyzerActionsForAnalyzer
+        {
+            private readonly DiagnosticAnalyzer _analyzer;
+            private readonly bool _analyzerActionsNeedFiltering;
+
+            private ImmutableDictionary<TLanguageKindEnum, ImmutableArray<SyntaxNodeAnalyzerAction<TLanguageKindEnum>>>? _lazyNodeActionsByKind;
+            private ImmutableDictionary<OperationKind, ImmutableArray<OperationAnalyzerAction>>? _lazyOperationActionsByKind;
```

Comment:
i wonder if there's a benefit here to using normal dict.

Reply:
Possibly, but accesses to this dictionary did not show up on the profile. The changes I have made in this PR are purely driven from the dictionary accesses that were showing up large CPU stacks in perfview runs across different tests.